# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [1]:
# Import your libraries:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score



# Challenge 1 - Explore the Scikit-Learn Datasets

Before starting to work on our own datasets, let's first explore the datasets that are included in this Python library. These datasets have been cleaned and formatted for use in ML algorithms.

First, we will load the diabetes dataset. Do this in the cell below by importing the datasets and then loading the dataset  to the `diabetes` variable using the `load_diabetes()` function ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_diabetes.html)).

In [2]:
# Your code here:
from sklearn.datasets import load_diabetes

diabetes = load_diabetes()

Let's explore this variable by looking at the different attributes (keys) of `diabetes`. Note that the `load_diabetes` function does not return dataframes. It returns you a Python dictionary.

In [3]:
# Your code here:
print("Keys in the diabetes dataset:", diabetes.keys())


Keys in the diabetes dataset: dict_keys(['data', 'target', 'frame', 'DESCR', 'feature_names', 'data_filename', 'target_filename', 'data_module'])


#### The next step is to read the description of the dataset. 

Print the description in the cell below using the `DESCR` attribute of the `diabetes` variable. Read the data description carefully to fully understand what each column represents.

*Hint: If your output is ill-formatted by displaying linebreaks as `\n`, it means you are not using the `print` function.*

In [4]:
# Your code here:
print(diabetes.DESCR)

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

  :Number of Instances: 442

  :Number of Attributes: First 10 columns are numeric predictive values

  :Target: Column 11 is a quantitative measure of disease progression one year after baseline

  :Attribute Information:
      - age     age in years
      - sex
      - bmi     body mass index
      - bp      average blood pressure
      - s1      tc, total serum cholesterol
      - s2      ldl, low-density lipoproteins
      - s3      hdl, high-density lipoproteins
      - s4      tch, total cholesterol / HDL
      - s5      ltg, possibly log of serum triglycerides level
      - s6      glu, blood sugar level

Note: Each of these 1

#### Based on the data description, answer the following questions:

1. How many attributes are there in the data? What do they mean?

1. What is the relation between `diabetes['data']` and `diabetes['target']`?

1. How many records are there in the data?

# Enter your answer here:
1. The diabetes dataset contains 10 numeric input attributes along with 1 numeric output attribute
2. The diabetes['data'] represents the feature matrix, which consists of the 10 input attributes for each patient. Each row corresponds to a single patient, and each column corresponds to one of the attributes.
The diabetes['target'] represents the output variable, which is the measure of disease progression for each patient. The relationship between diabetes['data'] and diabetes['target'] is that the input features (attributes) are used to predict the output variable (disease progression).
3. 442 records

#### Now explore what are contained in the *data* portion as well as the *target* portion of `diabetes`. 

Scikit-learn typically takes in 2D numpy arrays as input (though pandas dataframes are also accepted). Inspect the shape of `data` and `target`. Confirm they are consistent with the data description.

In [5]:
# Your code here:
data_shape = diabetes.data.shape
target_shape = diabetes.target.shape

print("Shape of data:", data_shape)
print("Shape of target:", target_shape)

Shape of data: (442, 10)
Shape of target: (442,)


# Challenge 2 - Perform Supervised Learning on the Dataset

The data have already been split to predictor (*data*) and response (*target*) variables. Given this information, we'll apply what we have previously learned about linear regression and apply the algorithm to the diabetes dataset.

#### Let's briefly revisit the linear regression formula:

```
y = β0 + β1X1 + β2X2 + ... + βnXn + ϵ
```

...where:

- X1-Xn: data 
- β0: intercept 
- β1-βn: coefficients 
- ϵ: error (cannot explained by model)
- y: target

Also take a look at the `sklearn.linear_model.LinearRegression` [documentation](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html).

#### In the cell below, import the `linear_model` class from `sklearn`. 

In [6]:
# Your code here:
from sklearn.linear_model import LinearRegression

#### Create a new instance of the linear regression model and assign the new instance to the variable `diabetes_model`.

In [7]:
# Your code here:
# Create a new instance of the linear regression model
diabetes_model = LinearRegression()

#### Next, let's split the training and test data.

Define `diabetes_data_train`, `diabetes_target_train`, `diabetes_data_test`, and `diabetes_target_test`. Use the last 20 records for the test data and the rest for the training data.

In [8]:
# Your code here:
# Get the data and target
X = diabetes.data
y = diabetes.target

# Define the training and test set sizes
train_size = len(X) - 20
test_size = 20

# Split the data into training and test sets
diabetes_data_train = X[:train_size]
diabetes_target_train = y[:train_size]

diabetes_data_test = X[-test_size:]
diabetes_target_test = y[-test_size:]

Fit the training data and target to `diabetes_model`. Print the *intercept* and *coefficients* of the model.

In [9]:
# Your code here:
# Fit the model to the training data
diabetes_model.fit(diabetes_data_train, diabetes_target_train)

# Print the intercept and coefficients of the model
print("Intercept:", diabetes_model.intercept_)
print("Coefficients:", diabetes_model.coef_)

Intercept: 152.76429169049118
Coefficients: [ 3.06094248e-01 -2.37635570e+02  5.10538048e+02  3.27729878e+02
 -8.14111926e+02  4.92799595e+02  1.02841240e+02  1.84603496e+02
  7.43509388e+02  7.60966464e+01]


#### Inspecting the results

From the outputs you should have seen:

- The intercept is a float number.
- The coefficients are an array containing 10 float numbers.

This is the linear regression model fitted to your training dataset.

#### Using your fitted linear regression model, predict the *y* of `diabetes_data_test`.

In [10]:
# Your code here:
# Predict the target values for the test dataset
predictions = diabetes_model.predict(diabetes_data_test)


#### Print your `diabetes_target_test` and compare with the prediction. 

In [11]:
# Your code here:
# Print the actual target values for the test dataset
print("Actual target values for diabetes_data_test:\n", diabetes_target_test)

# Print the predicted values for the test dataset
print("Predicted values for diabetes_data_test:\n", predictions)

Actual target values for diabetes_data_test:
 [233.  91. 111. 152. 120.  67. 310.  94. 183.  66. 173.  72.  49.  64.
  48. 178. 104. 132. 220.  57.]
Predicted values for diabetes_data_test:
 [197.61898486 155.44031962 172.88875144 111.53270645 164.79397301
 131.06765869 259.12441219 100.47873746 117.06005372 124.30261597
 218.36868146  61.19581944 132.24837933 120.33293546  52.54513009
 194.03746764 102.5756431  123.56778709 211.03465323  52.60221696]


#### Is `diabetes_target_test` exactly the same as the model prediction? Explain.

# Your explanation here:
It is not exactly the same. The model predictions are generated based on the linear regression algorithm and the predicted values is approximations of the actual target values, not exact matches.

# Challenge 3 - Peform Supervised Learning on a Pandas Dataframe

Now that we have dealt with data that has been formatted for scikit-learn, let's look at data that we will need to format ourselves.

In the next cell, load the `auto-mpg.csv` file included in this folder and assign it to a variable called `auto`.

In [12]:
# Your code here:
auto = pd.read_csv('/Users/anna/iron_hack/lab-supervised-learning-regression/auto-mpg.csv')

Look at the first 5 rows using the `head()` function:

In [13]:
# Your code here:
auto.head()

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   
1  15.0          8         350.0        165.0    3693          11.5   
2  18.0          8         318.0        150.0    3436          11.0   
3  16.0          8         304.0        150.0    3433          12.0   
4  17.0          8         302.0        140.0    3449          10.5   

   model_year                       car_name  
0          70  \t"chevrolet chevelle malibu"  
1          70          \t"buick skylark 320"  
2          70         \t"plymouth satellite"  
3          70              \t"amc rebel sst"  
4          70                \t"ford torino"

Evaluate the data to ensure that all numeric columns are correctly detected as such by pandas. If a column is misclassified as object, coerce it to numeric.

In [14]:
# Your code here:
auto.dtypes

mpg             float64
cylinders         int64
displacement    float64
horse_power     float64
weight            int64
acceleration    float64
model_year        int64
car_name         object
dtype: object

What is the newest model year and the oldest model year?

In [15]:
# Your code here:
oldest_model_year = auto['model_year'].min()
newest_model_year = auto['model_year'].max()

In [16]:
oldest_model_year

70

In [17]:
newest_model_year

82

Check the dataset for missing values and remove all rows containing at least one missing value.

In [18]:
# Your code here
auto.isnull().sum()

mpg             0
cylinders       0
displacement    0
horse_power     6
weight          0
acceleration    0
model_year      0
car_name        0
dtype: int64

In [19]:
auto = auto.dropna()


Find the frequency table for the `cylinders` column using the `value_counts()` function. How many possible values of cylinders are there?

In [20]:
# Your code here:
cylinder_counts = auto['cylinders'].value_counts()
cylinder_counts

4    199
8    103
6     83
3      4
5      3
Name: cylinders, dtype: int64

In [21]:
# Count the number of unique values in the 'cylinders' column
num_unique_cylinders = auto['cylinders'].nunique()
num_unique_cylinders

5

We would like to generate a linear regression model that will predict mpg. To do this, first drop the `car_name` column since it does not contain any quantitative data. Next separate the dataframe to predictor and response variables. Separate those into test and training data with 80% of the data in the training set and the remainder in the test set. 

Assign the predictor and response training data to `X_train` and `y_train` respectively. Similarly, assign the predictor and response test data to `X_test` and `y_test`.

*Hint: To separate data for training and test, use the `train_test_split` method we used in previous labs.*

In [22]:
# Your code here:
# Drop the 'car_name' column
auto = auto.drop('car_name', axis=1)

# Separate the features and target variables
X = auto.drop('mpg', axis=1) 
y = auto['mpg']               

# Split the data into training and test sets (80% training, 20% testing)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

Now we will processed and peform linear regression on this data to predict the mpg for each vehicle. 

#### In the next cell, create an instance of the linear regression model and call it `auto_model`. Fit `auto_model` with your training data.

In [23]:
# Your code here:
# Create an instance of the linear regression model
auto_model = LinearRegression()

# Fit the model to the training data
auto_model.fit(X_train, y_train)

LinearRegression()

# Challenge 4 - Evaluate the Model

In addition to evaluating your model with F-test and t-test, you can also use the *Coefficient of Determination* (a.k.a. *r squared score*). This method does not simply tell *yes* or *no* about the model fit but instead indicates how much variation can be explained by the model. Based on the r squared score, you can decide whether to improve your model in order to obtain a better fit.

You can learn about the r squared score [here](https://en.wikipedia.org/wiki/Coefficient_of_determination). In the end, we want the r-squared score to be as high as possible.

#### In the next cell, compute the predicted *y* based on `X_train` and call it `y_pred`. Then calcualte the r squared score between `y_pred` and `y_train` which indicates how well the estimated regression model fits the training data.

*Hint: r squared score can be calculated using `sklearn.metrics.r2_score` ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html)).*

In [24]:
# Your code here:
# Predict y based on X_train
y_pred = auto_model.predict(X_train)

# Calculate the R-squared score
r_squared = r2_score(y_train, y_pred)

# Print the R-squared score
print("R-squared score:", r_squared)

R-squared score: 0.8107227953093896


#### Our next step is to evaluate the model using the test data. 

We would like to ensure that our model is not overfitting the data. This means that our model was made to fit too closely to the training data by being overly complex. If a model is overfitted, it is not generalizable to data outside the training data. In that case, we need to reduce the complexity of the model by removing certain features (variables).

In the cell below, use the model to generate the predicted values for the test data and assign them to `y_test_pred`. Compute the r squared score of the predicted `y_test_pred` and the oberserved `y_test` data.

In [25]:
# Your code here:
# Predict y based on X_test
y_test_pred = auto_model.predict(X_test)

# Calculate the R-squared score for the test data
r_squared_test = r2_score(y_test, y_test_pred)

# Print the R-squared score for the test data
print("R-squared score for the test data:", r_squared_test)

R-squared score for the test data: 0.7942349075428586


#### Explaining the results

The r squared scores of the training data and the test data are pretty close (0.8146 vs 0.7818). This means our model is not overfitted. However, there is still room to improve the model fit. Move on to the next challenge.

# Challenge 5 - Improve the Model Fit

While the most common way to improve the fit of a model is by using [regularization](https://datanice.github.io/machine-learning-101-what-is-regularization-interactive.html), there are other simpler ways to improve model fit. The first is to create a simpler model. The second is to increase the train sample size.

Let us start with the easier option and increase our train sample size to 90% of the data. Create a new test train split and name the new predictors and response variables `X_train09`, `X_test09`, `y_train09`, `y_test09`.

In [26]:
# Your code here:
# Split the data into training and test sets (90% training, 10% testing)
X_train09, X_test09, y_train09, y_test09 = train_test_split(X, y, test_size=0.1, random_state=42)

Initialize a new linear regression model. Name this model `auto_model09`. Fit the model to the new sample (training) data.

In [27]:
# Your code here:
# Initialize a new linear regression model
auto_model09 = LinearRegression()

# Fit the model to the new training data
auto_model09.fit(X_train09, y_train09)

LinearRegression()

Compute the predicted values and r squared score for our new model and new sample data.

In [28]:
# Your code here:
# Predict values for the test data
y_test_pred = auto_model09.predict(X_test09)

# Calculate the R-squared score for the test data
r_squared_test_new = r2_score(y_test09, y_test_pred)

Compute the r squared score for the smaller test set. Is there an improvement in the test r squared?

In [29]:
# Your code here:
r_squared_test_new

0.8468911998183241

In [30]:
if r_squared_test_new > r_squared_test:
    print("Improvement in R-squared score with the new test set.")
else:
    print("No improvement in R-squared score with the new test set.")

Improvement in R-squared score with the new test set.


# Bonus Challenge - Backward Elimination 

The main way to produce a simpler linear regression model is to reduce the number of variables used in the model. In scikit-learn, we can do this by using recursive feature elimination. You can read more about RFE [here](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFE.html).

In the next cell, we will import RFE

In [36]:
from sklearn.feature_selection import RFE

Follow the documentation and initialize an RFE model using the `auto_model` linear regression model. Set `n_features_to_select=3`

In [37]:
# Your code here:
# Initialize the RFE model with the linear regression model
rfe = RFE(estimator=auto_model, n_features_to_select=3)

Fit the model and print the ranking

In [38]:
# Your code here:
rfe.fit(X_train, y_train)

RFE(estimator=LinearRegression(), n_features_to_select=3)

Feature importance is ranked from most important (1) to least important (4). Generate a model with the three most important features. The features correspond to variable names. For example, feature 1 is `cylinders` and feature 2 is `displacement`.

Perform a test-train split on this reduced column data and call the split data `X_train_reduced`, `X_test_reduced`, `y_test_reduced`, `y_train_reduced`. Use an 80% split.

In [39]:
# Your code here:
# Get the names of the selected features
selected_features = X.columns[rfe.support_]

# Print the selected features
print("Selected features:", selected_features.tolist())

# Create a new DataFrame with only the selected features
X_reduced = X[selected_features]

# Perform a train-test split on the reduced data (80% training, 20% testing)
X_train_reduced, X_test_reduced, y_train_reduced, y_test_reduced = train_test_split(X_reduced, y, test_size=0.2, random_state=42)

Selected features: ['cylinders', 'acceleration', 'model_year']


Generate a new model called `auto_model_reduced` and fit this model. Then proceed to compute the r squared score for the model. Did this cause an improvement in the r squared score?

In [40]:
# Your code here: 
# Initialize the new linear regression model for the reduced dataset
auto_model_reduced = LinearRegression()

# Fit the model to the reduced training data
auto_model_reduced.fit(X_train_reduced, y_train_reduced)

# Predict values for the reduced test data
y_test_pred_reduced = auto_model_reduced.predict(X_test_reduced)

# Calculate the R-squared score for the reduced model
r_squared_reduced = r2_score(y_test_reduced, y_test_pred_reduced)

# Print the R-squared score for the reduced model
print("R-squared score for the reduced model:", r_squared_reduced)

R-squared score for the reduced model: 0.7149033079956759


#### Conclusion

You may obtain the impression from this lab that without knowing statistical methods in depth, it is difficult to make major progress in machine learning. That is correct. If you are motivated to become a data scientist, statistics is the subject you must be proficient in and there is no shortcut. 

Completing these labs is not likely to make you a data scientist. But you will have a good sense about what are there in machine learning and what are good for you. In your future career, you can choose one of the three tracks:

* Data scientists who need to be proficient in statistical methods.

* Data engineers who need to be good at programming.

* Data integration specialists who are business or content experts but also understand data and programming. This cross-disciplinary track brings together data, technology, and business and will be in high demands in the next decade.